go/libraries/doltcore/sqle/tables.go

```diff
@@ -104,6 +108,147 @@ func (t *DoltTable) PartitionRows(ctx *sql.Context, _ sql.Partition) (sql.RowIte
 	return newRowIterator(t, ctx)
 }
 
+func (t *DoltTable) CreateIndex(ctx *sql.Context, indexName string, using sql.IndexUsing, constraint sql.IndexConstraint, columns []sql.IndexColumn, comment string) error {
+	if !doltdb.IsValidTableName(indexName) {
+		return fmt.Errorf("invalid index name `%s` as they must match the regular expression %s", indexName, doltdb.TableNameRegexStr)
+	}
+
+	// get the real column names as CREATE INDEX columns are case-insensitive
+	var realColNames []string
+	allTableCols := t.sch.GetAllCols()
+	for _, indexCol := range columns {
+		tableCol, ok := allTableCols.GetByName(indexCol.Name)
+		if !ok {
+			tableCol, ok = allTableCols.GetByNameCaseInsensitive(indexCol.Name)
+			if !ok {
+				return fmt.Errorf("column `%s` does not exist for the table", indexCol.Name)
+			}
+		}
+		realColNames = append(realColNames, tableCol.Name)
+	}
+
+	// create the index metadata, will error if index names are taken or an index with the same columns in the same order exists
+	index, err := t.sch.Indexes().AddIndexByColNames(indexName, realColNames, constraint == sql.IndexConstraint_Unique, comment)
+	if err != nil {
+		return err
+	}
+
+	// update the table schema with the new index
+	newSchemaVal, err := encoding.MarshalSchemaAsNomsValue(ctx, t.table.ValueReadWriter(), t.sch)
+	if err != nil {
+		return err
+	}
+	tableRowData, err := t.table.GetRowData(ctx)
+	if err != nil {
+		return err
+	}
+	indexData, err := t.table.GetIndexData(ctx)
+	if err != nil {
+		return err
+	}
+	newTable, err := doltdb.NewTable(ctx, t.table.ValueReadWriter(), newSchemaVal, tableRowData, indexData)
+	if err != nil {
+		return err
+	}
+
+	// populate the index row data
+	indexRowData, err := types.NewMap(ctx, newTable.ValueReadWriter())
```

Comment:
Shouldn't this call table.RebuildIndex?

Reply:
It should, I wrote this before writing `RebuildIndexData` and the related functions. I'll throw that in here instead, great catch!